### DISCUSSION TITLE
Should we write to disk every block?

### DISCUSSION THREAD

**Discussion by tac0turtle** [tendermint/tendermint#5784] [Ideas]
Yay, discussion boards. 🥳 

Currently, Tendermint writes to disk after every block. Writing to disk is intensive and doing it for each block adds up. 

What if we kept x amount of blocks and state updates in memory and write them all at once every x amount of blocks. This would allow for quick access of needed data for consensus. The downside here is if a node panics or falls over it will need to resync the blocks stored in memory. If the node is shutting down it can write all the blocks to disk prior to shut down. 

This process could lend itself to increases in performance but it would be good to benchmark beforehand.

**Comment by faddat** [comment]
So I am very interested in this, because when I'm working on embedded devices, the fact is I actually have all the RAM in the world, But cheap SD card will degrade with time, especially as there are more writes.

As I go forward, I have been thinking about doing stuff like putting the WAL file on a ram disk.

This is to ensure that we can run on the cheapest most basic devices.

**Comment by erikgrinaker** [comment]
We could just let the OS handle this for us. It will typically buffer writes in memory and only flush to disk when appropriate (e.g. when there is available IO capacity or the disk buffer fills up). If we made fsync behavior configurable and the node operator tuned some kernel parameters that might be sufficient.

**Reply by tac0turtle** [reply]
Peeking at goleveldb it seems they use fsync already. Is your suggestion adding another layer of it?

**Reply by erikgrinaker** [reply]
No, just making it configurable when we call it. E.g. now I think we fsync the WAL after every block write, we could let the operator disable that to improve throughput (at the expense of safety, since data will be lost if the OS crashes).

**Reply by tac0turtle** [reply]
Not sure this is configurable in some databases (I could be looking at the wrong place in goleveldb). I was thinking we have a cachedb and the cachedb write to the underlying db. The amount of blocks stored in cachdb would be configurable.

**Reply by erikgrinaker** [reply]
In goleveldb the caller specifies fsync behavior via `WriteOptions`: https://godoc.org/github.com/syndtr/goleveldb/leveldb/opt#WriteOptions

We expose this in tm-db as `Batch.WriteSync()`, as opposed to `Batch.Write()` which does not fsync.

The Linux kernel's disk buffer behavior can be configured via `vm` sysctls: https://lonesysadmin.net/2013/12/22/better-linux-disk-caching-performance-vm-dirty_ratio/

**Reply by tac0turtle** [reply]
Oh okay. Thank you for the links! need to read more on some of these concepts. 😄

**Reply by Spongebob-bear** [reply]
So is there any wal-light mode provided in current Tendermint?
